File: SnakeWear/Game.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

using Android.App;
using Android.Content;
using Android.Graphics;
using Android.OS;
using Android.Runtime;
using Android.Util;
using Android.Views;
using Android.Widget;
using Android.Graphics.Drawables;
using SnakeModel;

namespace SnakeWear
{
    public class Game : View
    {
        private SnakeGame game;
        private Renderer renderer;
        private System.Timers.Timer longPressTimer;
        private bool exit = false;

        public Game(Context context, int width, int height, int speed) : base(context)
        {
            game = new SnakeGame();

            renderer = new Renderer(game, width, height);

            //This is why I love lambdas. This is all the game steering I need to do.
            Touch += (s, e) => { if (e.Event.Action == MotionEventActions.Up) game.Goto(renderer.GetGameX((int)e.Event.GetX()), renderer.GetGameY((int)e.Event.GetY())); };
            Touch += (s, e) => { if (game.GameOver) GotoMenu(context); };
            ConfigureLongPress(context);

            var t = new System.Timers.Timer();
            t.Interval = 700/speed;
            t.Elapsed += (s, e) =>
            {
                game.Update();
                PostInvalidate();
                if (game.GameOver) t.Stop();
            };
            t.Start();
        }

        protected override void OnDraw(Canvas canvas)
        {
            base.OnDraw(canvas);
            renderer.Render(canvas);
        }

        private void GotoMenu(Context context)
        {
            MainActivity host = (MainActivity)context;
            host.ShowMainMenu();
        }

        private void ConfigureLongPress(Context context)
        {
            Touch += (s, e) =>
            {
                if (e.Event.Action == MotionEventActions.Down) longPressTimer.Start();
                if (e.Event.Action == MotionEventActions.Up)
                {
                    longPressTimer.Stop();
                    if (exit) GotoMenu(context);
                }
            };

            longPressTimer = new System.Timers.Timer();
            longPressTimer.Interval = 3000;
            longPressTimer.Elapsed += (s, e) => exit = true;
        }
    }   
}

File: android/snakewear/mobile/src/main/java/pontus/wearsnake/Game.java
package pontus.wearsnake;

import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.graphics.Canvas;
import android.os.Bundle;
import android.util.Log;
import android.view.KeyEvent;
import android.view.MotionEvent;
import android.view.View;

import com.google.android.gms.ads.AdListener;
import com.google.android.gms.ads.InterstitialAd;

import java.util.Random;
import java.util.Timer;
import java.util.TimerTask;

import pontus.wearsnake.model.SnakeGame;

import static android.content.Context.MODE_PRIVATE;

/**
 * Created by Pontus on 2018-01-05.
 */

public class Game extends View {
    private SnakeGame game;
    private Renderer renderer;
    private Timer timer;
    private InterstitialAd bigAd;
    private Random random;
    public Game(Context context) {
        super(context);
    }

    public Game(Context context, int width, int height, int speed, int highscore, InterstitialAd bigAd) {
        this(context);

        this.bigAd = bigAd;
        game = new SnakeGame();
        renderer = new Renderer(game, width, height, highscore);

        random = new Random();

        setFocusableInTouchMode(true);
        requestFocus();
        setOnKeyListener((view, keyCode, e) -> {
            if (keyCode == KeyEvent.KEYCODE_BACK && e.getRepeatCount() == 0) {
                gotoMenu(context);
                return true;
            }
            return super.onKeyDown(keyCode, e);
        });

        setOnTouchListener((s, e) -> {
            if (game.isGameOver()) {
                gotoMenu(context);
                return true;
            }

            if (e.getAction() == MotionEvent.ACTION_UP) {
                game.changeDirection(
                        renderer.getGameX((int) e.getX()),
                        renderer.getGameY((int) e.getY())
                );
            }

            s.performClick();
            return true;
        });

        timer = new Timer();
        TimerTask task = new TimerTask() {
            @Override
            public void run() {
                game.update();
                postInvalidate();
                if (game.isGameOver()) {
                    timer.cancel();
                }
            }
        };

        timer.schedule(task, 0, 700 / speed);
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        renderer.render(canvas);
    }

    private void gotoMenu(Context context) {
        timer.cancel();

        // Save highscore
        SharedPreferences prefs = context.getSharedPreferences(MainActivity.class.getSimpleName(), MODE_PRIVATE);
        int highscore = Math.max(prefs.getInt("highscore", 0), game.getScore());
        SharedPreferences.Editor edit = prefs.edit();
        edit.putInt("highscore", highscore);
        edit.apply();

        Intent goBack = new Intent(context, MainActivity.class);
        goBack.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);

        if (random.nextBoolean()) {
            bigAd.setAdListener(new AdListener() {
                @Override
                public void onAdClosed() {
                    context.startActivity(goBack);
                }
            });
            bigAd.show();
        } else {
            context.startActivity(goBack);
        }
    }
}
